src/vmm/src/lib.rs

```diff
@@ -370,6 +377,38 @@ impl Vmm {
     pub fn kvm_vm(&self) -> &Vm {
         &self.vm
     }
+
+    /// Saves the device states.
+    pub fn save_mmio_device_states(&mut self) -> Vec<ConnectedMmioDeviceState> {
+        let mut states: Vec<ConnectedMmioDeviceState> = Vec::new();
+        let device_manager = &mut self.mmio_device_manager;
+
+        for ((device_type, device_id), device_info) in device_manager.get_device_info().iter() {
+            let bus_device = device_manager
+                .get_device(device_type.clone(), device_id)
+                // Safe to unwrap() because we know the device exists.
+                .unwrap()
+                .lock()
+                .expect("Poisoned device lock");
+
+            let mmio_transport = bus_device
+                .as_any()
+                // Only MmioTransport implements BusDevice at this point.
+                .downcast_ref::<MmioTransport>()
+                .expect("Unexpected BusDevice type");
+
+            states.push(ConnectedMmioDeviceState {
+                mmio_device_state: MmioDeviceState::from_transport(&mmio_transport),
+                mmio_base: device_info.addr,
+                irq: device_info.irq,
+            });
+        }
+
+        // Sort the devices by mmio base addr since they will have to be added back in the same order.
+        states.sort_by(|a, b| a.mmio_base.partial_cmp(&b.mmio_base).unwrap());
```

Comment:
Ideally we shouldn't have to do this. Ordering is important for boot, but for restoring an already booted vm, we only care about restoring each device to its right address-space (mmio_base + len) and given its previous irq#. I don't see why the order would matter.

Reply:
You're right. It was useful if we wanted to reuse the `register_mmio_device` method, but we'll refactor the build path anyway.